docs/design/proposals/collections.md

```diff
@@ -0,0 +1,604 @@
+is equivalent to "for each item in expression, manage a resource". It's a
+declarative way to express iteration - you're asking KRO to iterate over a
+collection and create a resource for each element.
+
+```yaml
+forEach: ${expression} # CEL expression that evaluates to something iterable
+```
+
+The expression must evaluate to an iterable collection. When forEach is
+specified on a resource, KRO expands that single resource definition into
+multiple resources at runtime - one for each item in the collection. Each
+resource instance is tracked independently with its own reconciliation state.
+
+Example in an RGD:
+
+```yaml
+kind: ResourceGraphDefinition
+# ...
+spec:
+  resources:
+    # workers node in the DAG
+    - id: workers
+      forEach: ${range(1, 4)} # Creates 3 Pods
+      template:
+        apiVersion: v1
+        kind: Pod
+        metadata: { ... }
+        spec: { ... }
+```
+
+The `forEach` expression must evaluate to one of the following types:
+
+- **Arrays**: Any list value - `${[0, 1, 2]}`, `${spec.regions}`,
+  `${["us-east", "us-west"]}`, or `${pod.spec.containers}`
+- **Maps**: key-value pairs where iteration occurs over entries -
+  `${spec.labels}` or `${{"env": "prod", "tier": "web"}}`
+
+The expression can include CEL transformations or macros:
+`${spec.workers.filter(w, w.enabled)}` or `${spec.zones.map(z, z + "-a")}`.
+
+As the collection changes (items added/removed), KRO automatically reconciles
+the resources - creating new ones for added items and deleting resources for
+removed items. Each resource maintains a stable identity based on its position
+or key in the collection.
+
+### Iteration Context Variables
+
+When a resource uses forEach, special variables become available within CEL
+expressions in the template:
+
+- `each.item` - The current element being iterated over
+- `each.index` - The index (for arrays)
+- `each.key` - The key (for maps)
+- `each.value` - The value (for maps)
+- `each.length` - Total size of the collection
+
+Example usage:
+
+```yaml
+- id: workers
+  forEach: ${spec.workers} # ["alice", "bob", "charlie"]
+  template:
+    kind: Pod
+    metadata:
+      name: ${'worker-' + each.item} # worker-alice, worker-bob, worker-charlie
+      labels:
+        worker-name: ${each.item} # alice, bob, charlie
+        worker-index: ${string(each.index)} # "0", "1", "2"
+        total-workers: ${string(each.length)} # "3"
+```
+
+### `readyWhen` with Collections
+
+While `readyWhen` is an existing field in KRO, it gains new capabilities with
+collections. For resources using forEach, readyWhen can now define conditions
+across all instances in the collection:
+
+```yaml
+- id: workers
+  forEach: ${spec.workers}
+  # 'workers' node is ready when ALL worker pods are running
+  readyWhen: ${workers.all(w, w.status.phase == 'Running')}
+  template:
+    kind: Pod
+    # ...
+```
+
+The resource collection is considered ready only when ALL worker pods are
+running. You can also use:
```

Comment:
sugg: easier to read if all of the new aggregation functions are defined in a list. i.e. add `all` here.

Reply:
Or perhaps all of the new cel functions we're supporting. Right now it's a bit difficult to distinguish functions that already exist from net-new ones.

Reply:
Nice catch! Update the section to include all the supported functions/macros by default.